### PR TITLE
NSFS | Versioning | set_bucket_versioning enablement

### PR DESCRIPTION
**Pull request by romayalon**
Signed-off-by: Romy <romy2232@gmail.com>

### Explain the changes
FULL DOCS WILL BE ADDED AFTER THE DESIGN DISCUSSIONS WILL END

1. config
    1.1 NSFS_VERSIONING_ENABLED = false added so the feature won't be in use yet
3. namespace_fs 
    2.1 set_bucket_versioning implementation (reject on UNAUTHORIZED when no permissions to the bucket)
    2.2 addition of versioning mode to ctor
4. Object SDK -
    2.2 set_bucket_versioning -get bucket ns and try to set the new versioning mode if ns of type namespaceFS, else reject
    2.3 pass versioning mode on NamespaceFS object creation

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1. 

- [ ] Doc added/updated
- [ ] Tests added

### CODE REVIEW

File: src/sdk/namespace_fs.js [hunk 1]
```diff
@@ -180,15 +180,17 @@ class NamespaceFS {
      *  bucket_id: string;
      *  namespace_resource_id?: string;
      *  access_mode: string;
+     *  versioning: boolean;
```
Comment (jackyalbo) [inline src/sdk/namespace_fs.js:183]: Doesn't is suppose to string? one of those: 'DISABLED', 'SUSPENDED', 'ENABLED', maybe you should add versioning enum to nb.d.ts?

File: src/server/system_services/bucket_server.js [hunk 2]
```diff
@@ -589,7 +589,7 @@ function get_bucket_changes_versioning(req, bucket, update_request, single_bucke
     if (update_request.versioning === 'DISABLED') {
         throw new RpcError('BAD_REQUEST', 'Cannot set versioning to DISABLED');
     }
-    if (bucket.namespace) {
+    if (bucket.namespace && !bucket.namespace.write_resource.resource.nsfs_config.fs_root_path) {
         throw new RpcError('BAD_REQUEST', 'Cannot set versioning on namespace buckets');
```
Comment (jackyalbo) [inline src/server/system_services/bucket_server.js:593]: looks like you need to update the error, we will be able to set versioning on nsfs

File: src/test/unit_tests/test_bucketspace.js [hunk 3]
```diff
@@ -1193,3 +1192,138 @@ mocha.describe('nsfs account configurations', function() {
...
+
+    mocha.it('create account with uid, gid - default has permissions', async function() {
+        const { access_key, secret_key } = await generate_nsfs_account();
+        s3_uid6 = generate_s3_client(access_key, secret_key);
+    });
+
+    mocha.it('set bucket versioning - Enabled - should fail - no permissions', async function() {
+        try {
+            await s3_uid5.putBucketVersioning({ Bucket: bucket_name, VersioningConfiguration: { MFADelete: 'Disabled', Status: 'Enabled' } }).promise();
+            assert.fail(`put object succeeded for account without permissions`);
```
Comment (jackyalbo) [inline src/test/unit_tests/test_bucketspace.js:1310]:
```suggestion
            assert.fail(`put bucket versioning succeeded for account without permissions`);
```

File: src/test/unit_tests/test_bucketspace.js [hunk 4]
```diff
@@ -1193,3 +1192,138 @@ mocha.describe('nsfs account configurations', function() {
...
+        } catch (err) {
+            assert.ok(err.code === 'AccessDenied');
+        }
+    });
+
+    mocha.it('set bucket versioning - Enabled - admin - should fail - no permissions', async function() {
+        try {
+            await s3_admin.putBucketVersioning({ Bucket: bucket_name, VersioningConfiguration: { MFADelete: 'Disabled', Status: 'Enabled' } }).promise();
+            assert.fail(`put object succeeded for account without permissions`);
```
Comment (jackyalbo) [inline src/test/unit_tests/test_bucketspace.js:1319]: same as previous

File: src/test/unit_tests/test_bucketspace.js [hunk 5]
```diff
@@ -1193,3 +1192,138 @@ mocha.describe('nsfs account configurations', function() {
...
+        };
+        const { access_key, secret_key } = await generate_nsfs_account({ admin: true });
+        s3_admin = generate_s3_client(access_key, secret_key);
+        await s3_admin.putBucketPolicy({
+            Bucket: bucket_name,
+            Policy: JSON.stringify(policy)
+        }).promise();
+    });
+
+    mocha.it('create account uid, gid 5 - no permissions', async function() {
```
Comment (jackyalbo) [inline src/test/unit_tests/test_bucketspace.js:1297]: why is this a test case? looks like a test setup

File: src/test/unit_tests/test_bucketspace.js [hunk 6]
```diff
@@ -1193,3 +1192,138 @@ mocha.describe('nsfs account configurations', function() {
...
+        const { access_key, secret_key } = await generate_nsfs_account({ uid: 5, gid: 5 });
+        s3_uid5 = generate_s3_client(access_key, secret_key);
+    });
+
+    mocha.it('create account with uid, gid - default has permissions', async function() {
```
Comment (jackyalbo) [inline src/test/unit_tests/test_bucketspace.js:1302]: the same

File: src/test/unit_tests/test_bucketspace.js [hunk 7]
```diff
@@ -1193,3 +1192,138 @@ mocha.describe('nsfs account configurations', function() {
...
+        } catch (err) {
+            assert.ok(err.code === 'AccessDenied');
+        }
+    });
+
+    mocha.it('set bucket versioning - Enabled', async function() {
+        await s3_uid6.putBucketVersioning({ Bucket: bucket_name, VersioningConfiguration: { MFADelete: 'Disabled', Status: 'Enabled' } }).promise();
+    });
+
+});
```
Comment (jackyalbo) [inline src/test/unit_tests/test_bucketspace.js:1329]: This test in general starting to be very long - maybe its time to split to test_bucketspace_versionning?

**Review by jackyalbo — CHANGES_REQUESTED** [review]

File: src/sdk/namespace_fs.js (hunk 1 above)
Comment (romayalon) [reply src/sdk/namespace_fs.js]: it's a leftover, replaced it with the correct value

File: src/server/system_services/bucket_server.js (hunk 2 above)
Comment (romayalon) [reply src/server/system_services/bucket_server.js]: updated

File: src/test/unit_tests/test_bucketspace.js (hunk 5 above)
Comment (romayalon) [reply src/test/unit_tests/test_bucketspace.js]: moved it to the before section

File: src/test/unit_tests/test_bucketspace.js (hunk 6 above)
Comment (romayalon) [reply src/test/unit_tests/test_bucketspace.js]: same

File: src/test/unit_tests/test_bucketspace.js (hunk 7 above)
Comment (romayalon) [reply src/test/unit_tests/test_bucketspace.js]: moved it out to a new test file

**Review by jackyalbo — APPROVED** [review]

File: src/sdk/object_sdk.js [hunk 8]
```diff
@@ -869,6 +872,11 @@ class ObjectSDK {
     ///////////////////////
 
     async set_bucket_versioning(params) {
+        const ns = await this._get_bucket_namespace(params.name);
+        if (ns instanceof NamespaceFS) {
+            if (!config.NSFS_VERSIONING_ENABLED) throw new RpcError('BAD_REQUEST', 'Bad Request - nsfs versioning is unsupported');
+            await ns.set_bucket_versioning(params.versioning, this);
+        }
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/sdk/object_sdk.js]: I moved it to bucketspace_nb, thanks!

**Pull request merged**